src/monitord/monitor_actions.c

```diff
@@ -0,0 +1,256 @@
+    char str[OS_SIZE_1024 + 1];
+
+    memset(str, '\0', OS_SIZE_1024 + 1);
+    snprintf(str, OS_SIZE_1024 - 1, OS_AG_REMOVED, agent);
+
+    if (SendMSG(mond.a_queue, str, ARGV0, LOCALFILE_MQ) < 0) {
+        mond.a_queue = -1;  // set an invalid fd so we can attempt to reconnect later on.
+        mdebug1("Could not generate removed agent alert for '%s'", agent);
+        merror(QUEUE_SEND);
+    }
+}
+
+void monitor_send_disconnection_msg(char *agent) {
+    char str[OS_SIZE_1024 + 1];
+    int error;
+
+    memset(str, '\0', OS_SIZE_1024 + 1);
+    /* Send disconnected message */
+    snprintf(str, OS_SIZE_1024 - 1, AG_DISCON_MSG, agent);
+    if (error = mon_send_agent_msg(agent, str), error) {
+        if (error == 2) {
+            // Agent is no longer in the database
+            monitor_send_deletion_msg(agent);
+        } else {
+            mdebug1("Could not generate disconnected agent alert for '%s'", agent);
+        }
+    }
+}
+
+void monitor_agents_disconnection(){
+    int *agents_array;
+
+    agents_array = wdb_disconnect_agents(time(0) - mond.global.agents_disconnection_time, &sock);
+    if (agents_array) {
+        for (int i = 0; agents_array[i] != -1; i++) {
+            if (OSHash_Numeric_Add_ex(agents_to_alert_hash, agents_array[i], (void*)1) == 0) {
+                mdebug1("Can't add agent ID '%d' to the alerts hash table",agents_array[i]);
+            }
+        }
+        os_free(agents_array);
+    }
+}
+
+void monitor_agents_alert(){
+    unsigned int inode_it = 0;
+    OSHashNode *agent_hash_node = NULL;
+    OSHashNode *agent_hash_prev_node = NULL;
+
+    cJSON *j_agent_info = NULL;
+    cJSON *j_agent_status = NULL;
+    cJSON *j_agent_lastkeepalive = NULL;
+    cJSON *j_agent_name = NULL;
+    cJSON *j_agent_ip = NULL;
+
+    agent_hash_node = OSHash_Begin(agents_to_alert_hash, &inode_it);
+    while (agent_hash_node) {
+        j_agent_info = wdb_get_agent_info(atoi(agent_hash_node->key), &sock);
+        if (j_agent_info) {
+            j_agent_status = cJSON_GetObjectItem(j_agent_info->child, "connection_status");
+            j_agent_lastkeepalive = cJSON_GetObjectItem(j_agent_info->child, "last_keepalive");
+            j_agent_name = cJSON_GetObjectItem(j_agent_info->child, "name");
+            j_agent_ip = cJSON_GetObjectItem(j_agent_info->child, "register_ip");
+
+            if (cJSON_IsString(j_agent_status) && j_agent_status->valuestring != NULL &&
+                cJSON_IsString(j_agent_name) && j_agent_name->valuestring != NULL &&
+                cJSON_IsString(j_agent_ip) && j_agent_ip->valuestring != NULL &&
+                cJSON_IsNumber(j_agent_lastkeepalive)) {
+
+                    if (strcmp(j_agent_status->valuestring, "active") == 0) {
+                        /* The agent is now connected, removing from table */
+                        agent_hash_prev_node = agent_hash_node->prev;
+                        OSHash_Delete_ex(agents_to_alert_hash, agent_hash_node->key);
```

Comment:
I think we shouldn't use ex version of Has Table methods here.
Its safe to use the not secure methods because this is one-threaded. But more important: If this becomes multi-trheaded, the logic will not be safe neither as we are accesing elements as  agent_hash_node->prev without a lock. Thus, it can lead to misunderstood this logic will work as multithreaded.

Reply:
The __ex_ methods were replaced!